### PR TITLE
New Catalan Goodie

### PR DESCRIPTION
**Pull request by vadi95**
https://duck.co/ia/view/catalan

### CODE REVIEW

**Comment by vadi95** [comment]
https://duck.co/ia/view/catalan

**Comment by vadi95** [comment]
Hi! Can someone tell me with why the Travis CI build is failing? Thanks!

**Comment by GuiltyDolphin** [comment]
@vadi95 Add the link to the IA page in the PR comment.

File: lib/DDG/Goodie/Catalan.pm [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    s/^\s+//;
+    s/\s+$//;
+    return unless /^(?:what(?:'s| is) the )?(?<which>\d+)(?:th|rd|st)?(?: number)?(?: in the (?:series|sequence))?\??$/ && $1 <= 1470000;
+
+    my $n = $+{'which'};
+    my $suf = ordsuf($n);
+    my $ans = catalan($n);
+
+    return "The $n$suf catalan number is $ans.",
+      structured_answer => {
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/Catalan.pm]: How about using a template here? The `text` template should do nicely.

File: lib/DDG/Goodie/Catalan.pm [hunk 2]
```diff
@@ -0,0 +1,39 @@
+package DDG::Goodie::Catalan;
+
+# ABSTRACT: Returns the Nth Catalan number
+
+use DDG::Goodie;
+use strict;
+use ntheory qw(binomial);
+use Lingua::EN::Numbers::Ordinate qw(ordsuf);
+use bigint;
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/Catalan.pm]: Lets use the `NumberStyler` GoodieRole to format the input and output.

**Comment by moollaza** [comment]
@vadi95 sorry for the delay! We'll need to install the dependencies before I can deploy this to Beta!

**Comment by moollaza** [comment]
@vadi95 @GuiltyDolphin this is now deployed to Beta for testing!

File: lib/DDG/Goodie/Catalan.pm [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    return unless /^(?:what(?:'s| is) the )?(?<which>\d+)(?:th|rd|st)?(?: number)?(?: in the (?:series|sequence))?\??$/ && $1 <= 519;
+
+    my $n = $+{'which'};
+    my $style = number_style_for($n);
+    return unless $style;
+    
+    $n = $style->for_computation($n);
+    my $result = catalan($n);
+    if ($result >= (10**11)) {
+        $result = sprintf("%.10e", "$result");
```
Comment (moollaza) [inline lib/DDG/Goodie/Catalan.pm]: Can we use scientific notation here?

@GuiltyDolphin I forget, does the number style handle scientific notation?

Comment (GuiltyDolphin) [inline lib/DDG/Goodie/Catalan.pm]: @moollaza Uhuh. @vadi95 Anything to do with number formatting *should* go through NumberStyle (bar exceptional circumstances).

File: lib/DDG/Goodie/Catalan.pm [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  my $n = shift;
+  binomial(2*$n,$n)/($n+1);
+}
+
+
+# Handle statement
+handle remainder => sub {  
+    s/^\s+//;
+    s/\s+$//;
+    return unless /^(?:what(?:'s| is) the )?(?<which>\d+)(?:th|rd|st)?(?: number)?(?: in the (?:series|sequence))?\??$/ && $1 <= 519;
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/Catalan.pm]: Why 519?

File: t/Catalan.t [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+use DDG::Test::Goodie;
+
+zci answer_type => "catalan";
+zci is_cached   => 1;
+
+ddg_goodie_test(
+    [qw( DDG::Goodie::Catalan )],
+    'catalan 30' => test_zci(
+        'The 30th catalan number is 3.8149865021e+15.',
+        structured_answer => {
```
Comment (GuiltyDolphin) [inline t/Catalan.t]: Lets pull these out into a function for making the structured answer - makes it a lot easier to update later on.

File: lib/DDG/Goodie/Catalan.pm (hunk 3 above)
Comment (mintsoft) [inline lib/DDG/Goodie/Catalan.pm]: @moollaza Yup!

**Comment by jdorweiler** [comment]
This pull request is being closed due to inactivity but, if anyone would like to continue it, please @ mention 'duckduckhack-contributors' for help and http://docs.duckduckhack.com/submitting/submitting-overview.html for reference.

**Pull request closed without merging**